vm/devices/net/gdma/src/bnic.rs

```diff
@@ -346,42 +416,78 @@ impl BasicNic {
+                            vport.endpoint.stop().await;
+                        }
                     }
-                    Tristate::TRUE if !vport.task.is_running() => {
-                        if let (Some((sq_id, sq_cq_id)), Some((rq_id, rq_cq_id))) =
-                            (vport.queue_cfg.tx, vport.queue_cfg.rx)
-                        {
-                            let rx_packets = Arc::new(Default::default());
-
-                            let mut queues = vec![];
-                            vport
-                                .endpoint
-                                .get_queues(
-                                    vec![QueueConfig {
-                                        pool: Box::new(GuestBuffers {
-                                            gm: state.queues.gm.clone(),
-                                            rx_packets: Arc::clone(&rx_packets),
-                                            buffer_segments: Vec::new(),
-                                        }),
-                                        initial_rx: &[],
-                                        driver: Box::new(state.queues.driver.clone()),
-                                    }],
-                                    None,
-                                    &mut queues,
-                                )
-                                .await?;
-
-                            vport.task.insert(
+                    Tristate::TRUE => {
+                        let mut start_indices = Vec::new();
+                        let mut configs = Vec::new();
+                        let mut rx_packets_list: Vec<Arc<Mutex<Slab<RxPacket>>>> = Vec::new();
+
+                        for (idx, task) in vport.tasks.iter_mut().enumerate() {
+                            if task.task.is_running() {
+                                continue;
+                            }
+                            if task.queue_cfg.tx.is_some() && task.queue_cfg.rx.is_some() {
+                                // prepare per-task rx packet pool and QueueConfig
+                                let rx_packets = Arc::new(Default::default());
+                                rx_packets_list.push(Arc::clone(&rx_packets));
+                                configs.push(QueueConfig {
+                                    pool: Box::new(GuestBuffers {
+                                        gm: state.queues.gm.clone(),
+                                        rx_packets: Arc::clone(&rx_packets),
+                                        buffer_segments: Vec::new(),
+                                    }),
+                                    initial_rx: &[],
+                                    driver: Box::new(state.queues.driver.clone()),
+                                });
+                                start_indices.push(idx);
+                            }
+                        }
+
+                        if start_indices.is_empty() {
```

Comment:
Wouldn't this (i.e. start_indices empty) be also true in the case if all the tasks are already running? I think what we want to flag/catch here is the case where there is a task which doesnt has either Rx or Tx configured